core/lib/src/rocket.rs

```diff
@@ -703,54 +703,43 @@ impl Rocket {
         Ok(self)
     }
 
-    /// Starts the application server and begins listening for and dispatching
-    /// requests to mounted routes and catchers. Unless there is an error, this
-    /// function does not return and blocks until program termination.
-    ///
-    /// # Error
-    ///
-    /// If there is a problem starting the application, a [`LaunchError`] is
-    /// returned. Note that a value of type `LaunchError` panics if dropped
-    /// without first being inspected. See the [`LaunchError`] documentation for
-    /// more information.
+    /// Identical to `launch()`, but using a custom Tokio runtime. The runtime
+    /// has no other restrictions, and can have other tasks running on it.
     ///
     /// # Example
     ///
     /// ```rust
+    /// // This gives us the default behavior. Alternatively, we could use a
+    /// // `tokio::runtime::Builder` to configure with greater detail.
+    /// let runtime = tokio::runtime::Runtime::new();
+    ///
     /// # if false {
-    /// rocket::ignite().launch();
+    /// let fut = rocket::ignite().launch_on(&runtime);
+    /// fut.then(|_| {
+    ///     // do things after the server shuts down
+    /// });
     /// # }
     /// ```
-    pub fn launch(mut self) -> LaunchError {
+    pub async fn launch_on(mut self, runtime: &tokio::runtime::Runtime) -> Result<(), LaunchError> {
```

Comment:
`async` functions do nothing until the first `poll()`, so this would *require* the caller to do something with the returned future. I think this signature would be more honest if it was a non-`async` fn named `spawn_on` returning `Result<impl Future<Output = ()>, LaunchError>` (by returning an `async` block at the very end of the function).

Another reason for this is error handling: AFAICT `handle.await` can return non-launch `Error`s, but this would report them as `LaunchError`.

Reply:
As you mentioned, `handle.await` can return other errors. Should the `Future` output a `Result` as well? The only other option would be ignoring it or panicking.

Reply:
I suppose that should return a `Result` as well, yes. That would be necessary for a use case like "log a message and restart the server after a failure"

Reply:
Done in recent commit.